File: src/main/java/ch/admin/seco/jobroom/model/JobOffer.java
package ch.admin.seco.jobroom.model;


import com.fasterxml.jackson.annotation.JsonIgnore;
import org.joda.time.DateTime;
import org.joda.time.LocalDate;

import javax.persistence.*;
import javax.validation.Valid;
import javax.validation.constraints.NotNull;
import java.sql.Timestamp;

@Entity
public class JobOffer {

    @Id
    @GeneratedValue(generator="joboffer_gen")
    @SequenceGenerator(name="joboffer_gen", sequenceName="aoste_seq")
    private Integer id;

    @Version
    @JsonIgnore
    private Integer version;

    @NotNull
    private LocalDate publicationStartDate;

    private LocalDate publicationEndDate;

    private String reference;

    private String url;

    @Embedded
    @Valid
    @NotNull
    private Job job;

    @Embedded
    @NotNull
    private Company company;

    @Embedded
    @NotNull
    private Contact contact;

    @Embedded
    @NotNull
    private Application application;

    @ManyToOne
    @JoinColumn(name="owner_id")
    @JsonIgnore
    private RestAccessKey owner;

    @JsonIgnore
    private Timestamp creationDate;

    @JsonIgnore
    private Timestamp lastModificationDate;

    @JsonIgnore
    private Timestamp cancellationDate;

    @JsonIgnore
    private String cancellationReasonCode;

    @PrePersist
    void onCreate() {
        creationDate = new Timestamp(DateTime.now().getMillis());
    }

    @PreUpdate
    void onUpdate() {
        lastModificationDate = new Timestamp(DateTime.now().getMillis());
    }

    public JobOffer() {
    }

    public JobOffer(Integer id, Integer version,
                    LocalDate publicationStartDate, LocalDate publicationEndDate,
                    String reference, String url,
                    Job job, Company company, Contact contact, Application application,
                    RestAccessKey owner, Timestamp creationDate, Timestamp lastModificationDate,
                    Timestamp cancellationDate, String cancellationReasonCode) {
        this.version = version;
        this.publicationStartDate = publicationStartDate;
        this.publicationEndDate = publicationEndDate;
        this.reference = reference;
        this.url = url;
        this.job = job;
        this.company = company;
        this.contact = contact;
        this.application = application;
        this.owner = owner;
        this.creationDate = creationDate;
        this.lastModificationDate = lastModificationDate;
        this.cancellationDate = cancellationDate;
        this.cancellationReasonCode = cancellationReasonCode;
    }

    public Integer getId() {
        return id;
    }

    public void setId(Integer id) {
        this.id = id;
    }

    public Integer getVersion() {
        return version;
    }

    public void setVersion(Integer version) {
        this.version = version;
    }

    public LocalDate getPublicationStartDate() {
        return publicationStartDate;
    }

    public void setPublicationStartDate(LocalDate publicationStartDate) {
        this.publicationStartDate = publicationStartDate;
    }

    public LocalDate getPublicationEndDate() {
        return publicationEndDate;
    }

    public void setPublicationEndDate(LocalDate publicationEndDate) {
        this.publicationEndDate = publicationEndDate;
    }

    public String getReference() {
        return reference;
    }

    public void setReference(String reference) {
        this.reference = reference;
    }

    public String getUrl() {
        return url;
    }

    public void setUrl(String url) {
        this.url = url;
    }

    public Job getJob() {
        return job;
    }

    public void setJob(Job job) {
        this.job = job;
    }

    public Company getCompany() {
        return company;
    }

    public void setCompany(Company company) {
        this.company = company;
    }

    public Contact getContact() {
        return contact;
    }

    public void setContact(Contact contact) {
        this.contact = contact;
    }

    public Application getApplication() {
        return application;
    }

    public void setApplication(Application application) {
        this.application = application;
    }

    public RestAccessKey getOwner() {
        return owner;
    }

    public void setOwner(RestAccessKey owner) {
        this.owner = owner;
    }

    public Timestamp getCreationDate() {
        return creationDate;
    }

    public void setCreationDate(Timestamp creationDate) {
        this.creationDate = creationDate;
    }

    public Timestamp getLastModificationDate() {
        return lastModificationDate;
    }

    public void setLastModificationDate(Timestamp lastModificationDate) {
        this.lastModificationDate = lastModificationDate;
    }

    public Timestamp getCancellationDate() {
        return cancellationDate;
    }

    public void setCancellationDate(Timestamp cancellationDate) {
        this.cancellationDate = cancellationDate;
    }

    public String getCancellationReasonCode() {
        return cancellationReasonCode;
    }

    public void setCancellationReasonCode(String cancellationReasonCode) {
        this.cancellationReasonCode = cancellationReasonCode;
    }
}

File: src/main/java/ch/admin/seco/jobroom/model/JobOffer.kt
package ch.admin.seco.jobroom.model

import com.fasterxml.jackson.annotation.JsonIgnore
import org.joda.time.DateTime
import org.joda.time.LocalDate
import java.sql.Timestamp
import javax.persistence.*

@Entity data class JobOffer(

        @Id
        @GeneratedValue(generator="joboffer_gen")
        @SequenceGenerator(name="joboffer_gen", sequenceName="aoste_seq")
        var id: Int?,

        @Version
        var version: Int?,

        val publicationStartDate: LocalDate,

        var publicationEndDate: LocalDate?,

        var reference: String?,

        var url: String?,

        @Embedded
        @get:javax.validation.Valid
        val job: Job,

        @Embedded
        val company: Company,

        @Embedded
        val contact: Contact,

        @Embedded
        val application: Application,

        @ManyToOne
        @JoinColumn(name="owner_id")
        @get:JsonIgnore
        var owner: RestAccessKey?,

        @get:JsonIgnore
        var creationDate: Timestamp?,

        @get:JsonIgnore
        var lastModificationDate: Timestamp?,

        @get:JsonIgnore
        var cancellationDate: Timestamp?,

        @get:JsonIgnore
        var cancellationReasonCode: String?
) {
    // This default constructor is required by JPA
    constructor() : this(null,  null, LocalDate.now(), null, null, null, Job(), Company(), Contact(), Application(), null, null, null, null, null)

    @PrePersist
    fun onCreate() {
        creationDate = Timestamp(DateTime.now().millis)
    }

    @PreUpdate
    fun onUpdate() {
        lastModificationDate = Timestamp(DateTime.now().millis)
    }
}

@Embeddable
data class Job(

        val title: String,

        @Lob
        @field:javax.validation.constraints.Size(max = 10000)
        val description: String,

        val workingTimePercentageFrom: Int,
        val workingTimePercentageTo: Int,
        var startDate: LocalDate?,
        var endDate: LocalDate?,

        @Embedded
        val location: Location,

        @field:javax.validation.constraints.Size(max=5)
        @ElementCollection(fetch = FetchType.EAGER)
        var languageSkills: Collection<LanguageSkill>?

) {
    // This default constructor is required by JPA
    constructor() : this("", "", 0, 0, null, null, Location(), null)
}

@Embeddable
data class LanguageSkill(

        // Hibernate bug : @ElementCollection is incompatible with DefaultComponentSafeNamingStrategy property
        // Need to explicitly define column names
        @Column(name="language")
        val language: String,

        @Column(name="spokenlevel")
        @Enumerated(EnumType.STRING)
        val spokenLevel: LanguageSkill.Level,

        @Column(name="writtenlevel")
        @Enumerated(EnumType.STRING)
        val writtenLevel: LanguageSkill.Level
) {
    enum class Level {
        no_knowledge,
        basic_knowledge,
        good,
        very_good
    }

    // This default constructor is required by JPA
    constructor() : this("", Level.good, Level.good)
}

@Embeddable
data class Location(

        val countryCode: String,
        var locality: String?,
        var postalCode: String?,
        var additionalDetails: String?
) {
    // This default constructor is required by JPA
    constructor() : this("", null, null, null)
}

@Embeddable
data class Company (

        val name: String,
        val countryCode: String,
        val street: String,
        val houseNumber: String,
        val locality: String,
        val postalCode: String,
        var phoneNumber: String?,
        var email: String?,
        var website: String?,

        @Embedded
        var postbox: Postbox?
){
    // This default constructor is required by JPA
    constructor() : this("", "", "", "", "", "", null, null, null, null)
}


@Embeddable
data class Postbox(

        var number: String?,
        var locality: String?,
        var postalCode: String?
) {
    // This default constructor is required by JPA
    constructor() : this(null, null, null)
}


@Embeddable
data class Contact(

        @Enumerated(EnumType.STRING)
        val title: Contact.Title,

        val firstName: String,
        val lastName: String,
        val phoneNumber: String,
        val email: String
) {
    enum class Title {
        mister,
        madam
    }

    // This default constructor is required by JPA
    constructor() : this(Title.mister, "", "", "", "")
}

@Embeddable
data class Application(

        val telephonic: Boolean,
        val written: Boolean,
        val electronic: Boolean,
        var additionalDetails: String?,
        var url: String?
) {
    // This default constructor is required by JPA
    constructor() : this(false, false, false, null, null)
}
